Support for receiving ID via TraceParent HTTP header.

vancem opened the pull request:
Also moved the fetching of ANY HTTP header fields until AFTER the most stringent
check to see if logging is on (thus deferring work if the logger is sampling).

This sets things up for W3C support (traceparent HTTP header field) and allows for efficient sampling (since we don't fetch the ID string until AFTER we have decided to log the event. 

@davidfowl  @glennc  @lmolkova

I will followup via e-mail on what else is needed for this end-to-end.

Commit:
Support for receiving ID via TraceParent HTTP header.

Also moved the fetching of ANY HTTP header fields until AFTER the most stringent
check to see if logging is on (thus deferring work if the logger is sampling).

davidfowl commented:
@vancem You'll want to add to these tests here https://github.com/aspnet/AspNetCore/blob/425c196cba530b161b120a57af8f1dd513b96f67/src/Hosting/Hosting/test/HostingApplicationTests.cs

lmolkova commented on src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:
```diff
@@ -240,9 +238,13 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
         {
             var activity = new Activity(ActivityName);
+
+            if (!httpContext.Request.Headers.TryGetValue(TraceParentHeaderName, out var requestId))
```
Could you please also read the tracestate if traceparent is present?

vancem replied:
Done

Commit:
Merge remote-tracking branch 'upstream/master' into traceparentSupport

Commit:
Add TraceState and tests

Added transfer of TraceState from http header to Activity
Added tests to insure that we were reading traceparent and tracestate headers as expected.

vancem commented:
@davidfowl - I have added a basic test that shows that we correctly pull off the traceparent and tracestate HTTP headers.   

I think I have addressed all feedback.  Let me know if there is anything else to be done.

JamesNK commented on src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:
```diff
@@ -73,6 +68,10 @@ public void BeginRequest(HttpContext httpContext, ref HostingApplication.Context
             {
+                // Get the request ID (first try TraceParent header otherwise Request-ID header
+                if (!httpContext.Request.Headers.TryGetValue(TraceParentHeaderName, out var correlationId))
+                    httpContext.Request.Headers.TryGetValue(RequestIdHeaderName, out correlationId);
```
ASP.NET style guide is to always include { } for single line if statements

JamesNK commented on src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:
```diff
@@ -240,12 +239,18 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
             var activity = new Activity(ActivityName);
+
+            if (!httpContext.Request.Headers.TryGetValue(TraceParentHeaderName, out var requestId))
+                httpContext.Request.Headers.TryGetValue(RequestIdHeaderName, out requestId);
```
Add { }

JamesNK commented on src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:
```diff
@@ -240,12 +239,18 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
             {
                 activity.SetParentId(requestId);
+                if (httpContext.Request.Headers.TryGetValue(TraceStateHeaderName, out var traceState))
+                    activity.TraceStateString = traceState;
```
Add { }

JamesNK reviewed (commented):
Style stuff

Commit:
Review feedback, added {}

davidfowl reviewed (approved):
LGTM

davidfowl commented:
Thanks!
